metrics.go

```diff
@@ -0,0 +1,416 @@
+		"megabyte",
+	}
+}
+
+func MebiByte() MetricUnit {
+	return MetricUnit{
+		"mebibyte",
+	}
+}
+
+func GigaByte() MetricUnit {
+	return MetricUnit{
+		"gigabyte",
+	}
+}
+
+func GibiByte() MetricUnit {
+	return MetricUnit{
+		"gibibyte",
+	}
+}
+
+func TeraByte() MetricUnit {
+	return MetricUnit{
+		"terabyte",
+	}
+}
+
+func TebiByte() MetricUnit {
+	return MetricUnit{
+		"tebibyte",
+	}
+}
+
+func PetaByte() MetricUnit {
+	return MetricUnit{
+		"petabyte",
+	}
+}
+
+func PebiByte() MetricUnit {
+	return MetricUnit{
+		"pebibyte",
+	}
+}
+
+func ExaByte() MetricUnit {
+	return MetricUnit{
+		"exabyte",
+	}
+}
+
+func ExbiByte() MetricUnit {
+	return MetricUnit{
+		"exbibyte",
+	}
+}
+
+func Ratio() MetricUnit {
+	return MetricUnit{
+		"ratio",
+	}
+}
+
+func Percent() MetricUnit {
+	return MetricUnit{
+		"percent",
+	}
+}
+
+func CustomUnit(unit string) MetricUnit {
+	re := regexp.MustCompile(`[^a-z]+`)
+	return MetricUnit{
+		re.ReplaceAllString(unit, ""),
+	}
+}
+
+type Metric interface {
+	GetType() string
+	GetTags() map[string]string
+	GetKey() string
+	GetUnit() string
+	GetTimestamp() int
+	SerializeValue() string
+	SerializeTags() string
+}
+
+type abstractMetric struct {
+	key       string
+	unit      MetricUnit
+	tags      map[string]string
+	timestamp int
+}
+
+func (am abstractMetric) GetTags() map[string]string {
+	return am.tags
+}
+
+func (am abstractMetric) GetKey() string {
+	return am.key
+}
+
+func (am abstractMetric) GetUnit() string {
+	return am.unit.toString()
+}
+
+func (am abstractMetric) GetTimestamp() int {
+	return am.timestamp
+}
+
+func (am abstractMetric) SerializeTags() string {
+	var sb strings.Builder
+
+	values := make([]string, 0, len(am.tags))
+	for k := range am.tags {
+		values = append(values, k)
+	}
+	sortSlice(values)
+
+	for _, key := range values {
+		val := sanitizeValue(am.tags[key])
+		key = sanitizeKey(key)
+		sb.WriteString(fmt.Sprintf("%s:%s,", key, val))
+	}
+	s := sb.String()
+	if len(s) > 0 {
+		s = s[:len(s)-1]
+	}
+	return s
+}
+
+// Counter Metric.
+type CounterMetric struct {
+	value float64
+	abstractMetric
+}
+
+func (c *CounterMetric) Add(value float64) {
+	c.value += value
+}
+
+func (c CounterMetric) GetType() string {
+	return "c"
+}
+
+func (c CounterMetric) SerializeValue() string {
+	return fmt.Sprintf(":%v", c.value)
+}
+
+func NewCounterMetric(key string, unit MetricUnit, tags map[string]string, timestamp int, value float64) CounterMetric {
+	am := abstractMetric{
+		key,
+		unit,
+		tags,
+		timestamp,
+	}
+
+	return CounterMetric{
+		value,
+		am,
+	}
+}
+
+// Distribution Metric.
+type DistributionMetric struct {
+	values []float64
+	abstractMetric
+}
+
+func (d *DistributionMetric) Add(value float64) {
+	d.values = append(d.values, value)
+}
+
+func (d DistributionMetric) GetType() string {
+	return "d"
+}
+
+func (d DistributionMetric) SerializeValue() string {
+	var sb strings.Builder
+	for _, el := range d.values {
+		sb.WriteString(fmt.Sprintf(":%v", el))
```

Comment:
Why uses `%v` when we already know that `d.values` is a float64 slice?

Reply:
because this way, when serializing to a string, a float of `1` get serialized to  `1` instead of `1.000000`

Reply:
@viglia Mind to put a comment on the code about it? That would be good for people reading the code in the future